File: client/src/services/auditService.ts
import axios from 'axios';
import { extractErrorMessage } from '../utils/errorHandler';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8083';

// Create axios instance with auth interceptor
const auditApi = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add request interceptor to include JWT token
auditApi.interceptors.request.use((config) => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

export interface AuditLog {
  id: number;
  userId: number | null;
  userEmail: string;
  userRole: string;
  clientId: number | null;
  entityType: string;
  entityId: number | null;
  action: string;
  actionDetails: string;
  oldValues: string | null;
  newValues: string | null;
  ipAddress: string | null;
  userAgent: string | null;
  sessionId: string | null;
  requestId: string | null;
  status: string;
  failureReason: string | null;
  createdAt: string;
}

export interface AuditStatistics {
  totalEvents: number;
  securityEvents: number;
  failedEvents: number;
  successRate: number;
  startDate: string;
  endDate: string;
  generatedAt: string;
}

export interface AuditSearchFilters {
  searchTerm?: string;
  entityType?: string;
  action?: string;
  startDate?: string;
  endDate?: string;
  page?: number;
  size?: number;
}

export class AuditService {
  // Get all audit logs with pagination
  async getAllAuditLogs(page: number = 0, size: number = 20): Promise<{ content: AuditLog[]; totalElements: number; totalPages: number }> {
    try {
      const response = await auditApi.get(`/api/v1/audit/logs?page=${page}&size=${size}`);
      return {
        content: response.data.content,
        totalElements: response.data.totalElements,
        totalPages: response.data.totalPages
      };
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Search audit logs with filters
  async searchAuditLogs(filters: AuditSearchFilters): Promise<{ content: AuditLog[]; totalElements: number; totalPages: number }> {
    try {
      const params = new URLSearchParams();
      if (filters.searchTerm) params.append('searchTerm', filters.searchTerm);
      if (filters.entityType) params.append('entityType', filters.entityType);
      if (filters.action) params.append('action', filters.action);
      if (filters.startDate) params.append('startDate', filters.startDate);
      if (filters.endDate) params.append('endDate', filters.endDate);
      if (filters.page !== undefined) params.append('page', filters.page.toString());
      if (filters.size !== undefined) params.append('size', filters.size.toString());

      const response = await auditApi.get(`/api/v1/audit/logs/search?${params.toString()}`);
      return {
        content: response.data.content,
        totalElements: response.data.totalElements,
        totalPages: response.data.totalPages
      };
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Get recent audit logs
  async getRecentAuditLogs(): Promise<AuditLog[]> {
    try {
      const response = await auditApi.get('/api/v1/audit/logs/recent');
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Get security audit logs
  async getSecurityAuditLogs(page: number = 0, size: number = 20): Promise<{ content: AuditLog[]; totalElements: number; totalPages: number }> {
    try {
      const response = await auditApi.get(`/api/v1/audit/logs/security?page=${page}&size=${size}`);
      return {
        content: response.data.content,
        totalElements: response.data.totalElements,
        totalPages: response.data.totalPages
      };
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Get failed audit logs
  async getFailedAuditLogs(page: number = 0, size: number = 20): Promise<{ content: AuditLog[]; totalElements: number; totalPages: number }> {
    try {
      const response = await auditApi.get(`/api/v1/audit/logs/failed?page=${page}&size=${size}`);
      return {
        content: response.data.content,
        totalElements: response.data.totalElements,
        totalPages: response.data.totalPages
      };
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Get audit statistics
  async getAuditStatistics(startDate?: string, endDate?: string): Promise<AuditStatistics> {
    try {
      const params = new URLSearchParams();
      if (startDate) params.append('startDate', startDate);
      if (endDate) params.append('endDate', endDate);

      const response = await auditApi.get(`/api/v1/audit/statistics/summary?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Export audit logs to CSV
  async exportAuditLogsToCsv(startDate: string, endDate: string): Promise<Blob> {
    try {
      const response = await auditApi.get(`/api/v1/audit/export/csv?startDate=${startDate}&endDate=${endDate}`, {
        responseType: 'blob'
      });
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Generate compliance report
  async generateComplianceReport(startDate: string, endDate: string): Promise<string> {
    try {
      const response = await auditApi.get(`/api/v1/audit/report/compliance?startDate=${startDate}&endDate=${endDate}`);
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Generate daily summary report
  async generateDailySummaryReport(): Promise<string> {
    try {
      const endDate = new Date().toISOString();
      const startDate = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(); // Last 24 hours
      const response = await auditApi.get(`/api/v1/audit/report/compliance?startDate=${startDate}&endDate=${endDate}`);
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }

  // Generate security events report
  async generateSecurityEventsReport(): Promise<string> {
    try {
      const endDate = new Date().toISOString();
      const startDate = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString(); // Last 30 days
      const response = await auditApi.get(`/api/v1/audit/report/compliance?startDate=${startDate}&endDate=${endDate}`);
      return response.data;
    } catch (error: any) {
      throw new Error(extractErrorMessage(error));
    }
  }
}

export const auditService = new AuditService();

File: services/banking-service/src/main/java/com/microbank/bankingservice/service/AuditService.java
package com.microbank.bankingservice.service;

import com.microbank.bankingservice.domain.AuditAction;
import com.microbank.bankingservice.domain.AuditEntityType;
import com.microbank.bankingservice.domain.AuditLog;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;

public interface AuditService {
    
    /**
     * Log an audit event
     */
    AuditLog logEvent(Long userId, String userEmail, String userRole, 
                      AuditEntityType entityType, Long entityId, 
                      AuditAction action, String actionDetails);
    
    /**
     * Log an audit event with change details
     */
    AuditLog logEventWithChanges(Long userId, String userEmail, String userRole,
                                AuditEntityType entityType, Long entityId,
                                AuditAction action, String actionDetails,
                                String oldValues, String newValues);
    
    /**
     * Log a security event
     */
    AuditLog logSecurityEvent(String userEmail, String ipAddress, 
                             AuditAction action, String details);
    
    /**
     * Log a system event
     */
    AuditLog logSystemEvent(AuditAction action, String details);
    
    /**
     * Get audit logs for a specific entity
     */
    List<AuditLog> getAuditLogsForEntity(AuditEntityType entityType, Long entityId);
    
    /**
     * Get audit logs for a specific user
     */
    List<AuditLog> getAuditLogsForUser(Long userId);
    
    /**
     * Get audit logs for a specific client
     */
    List<AuditLog> getAuditLogsForClient(Long clientId);
    
    /**
     * Get audit logs by action type
     */
    List<AuditLog> getAuditLogsByAction(AuditAction action);
    
    /**
     * Get audit logs within a date range
     */
    List<AuditLog> getAuditLogsByDateRange(LocalDateTime startDate, LocalDateTime endDate);
    
    /**
     * Get all audit logs with pagination
     */
    Page<AuditLog> getAllAuditLogs(Pageable pageable);
    
    /**
     * Search audit logs with filters
     */
    Page<AuditLog> searchAuditLogs(String searchTerm, AuditEntityType entityType, 
                                   AuditAction action, LocalDateTime startDate, 
                                   LocalDateTime endDate, Pageable pageable);
    
    /**
     * Generate audit report for compliance
     */
    String generateComplianceReport(LocalDateTime startDate, LocalDateTime endDate);
    
    /**
     * Export audit logs to CSV
     */
    byte[] exportAuditLogsToCsv(LocalDateTime startDate, LocalDateTime endDate);
    
    /**
     * Clean up old audit logs (retention policy) - default 365 days
     */
    void cleanupOldAuditLogs();
    
    /**
     * Clean up old audit logs with custom retention period
     */
    void cleanupOldAuditLogs(int retentionDays);
}
